### PR TITLE
Emit warnings for integer constant overflow in implicit conversions

### LINKED ISSUE #8298: Emit warnings for overflows of function and generic arguments
While implementing warnings for overflows of types declarations in https://github.com/shader-slang/slang/pull/8281 I discovered that we do not emit warnings for all cases of overflows for function or generic arguments.

For example, the following code does not emit a warning:
void functon(int x) { ... }
...
function(UINT64_MAX)

We should add logic to handle cases like this, probably by emitting the warnings somewhere inside _coerce.
Ideally this should also handle warnings for types in mathematical expressions (but not bitwise operations).

### PR DESCRIPTION
**Pull request by jvepsalainen-nv**
Add a new `integer-constant-overflow` warning diagnostic that fires when a compile-time constant integer value does not fit in the target type during implicit conversions. This covers function arguments, generic value arguments, and variable initializers.

The check uses `tryFoldIntegerConstantExpression` to evaluate constant expressions and `getIntValueBitSize` / `getMaximumTypeBitSize` to detect when the value needs more bits than the target type provides. It runs at all conversion cost levels below `kConversionCost_Explicit`, catching cases that the existing `UnrecommendedImplicitConversion` warning missed because `getImplicitConversionCostWithKnownArg` reduced the cost below the warning threshold.

The check is deliberately bit-width-based (not range-aware) to avoid false positives on common shader patterns where hex constants serve as bit patterns in same-width signed types (e.g. `int x = 0xFF030206` for packed bytes). This means values like `128 -> int8_t` that are outside the signed range but fit in the bit width will not trigger the warning.

Tests added in `tests/diagnostics/integer-constant-overflow.slang`:
- Function arguments that overflow (int8_t, uint8_t, int16_t, int, uint)
- Generic value arguments that overflow
- Variable initializer overflow
- Constant expression overflow (e.g. 100 + 200 to int8_t)
- Bit-width boundary cases (e.g. 4294967296 to uint)
- No false positives for in-range values, same-width sign reinterpretation, -1 to unsigned, or explicit casts

Fixes #8298

### CODE REVIEW

File: tests/diagnostics/integer-constant-overflow.slang [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    // Function arguments that fit - should NOT warn
+    funcInt8(127);
+    funcInt8(-128);
+    funcUint8(255);
+    funcInt16(32767);
+    funcInt16(-32768);
+    funcUint16(65535);
+    funcInt(42);
+    funcUint(42u);
+    funcHalf(2048);
```
Comment (skiminki-nv) [inline tests/diagnostics/integer-constant-overflow.slang:101]: I'd use a larger value for the half-float. Something that is beyond 16 bit integers but within 16-bit half-float. Something like 131072, perhaps?

File: tests/diagnostics/integer-constant-overflow.slang [hunk 2]
```diff
@@ -0,0 +1,98 @@
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):
+
+void funcInt8(int8_t x) {}
+void funcInt16(int16_t x) {}
+void funcUint16(uint16_t x) {}
+void funcInt(int x) {}
+void funcUint(uint x) {}
+void funcUint8(uint8_t x) {}
+void funcHalf(half x) {}
```
Comment (skiminki-nv) [inline tests/diagnostics/integer-constant-overflow.slang:11]: I think we should add funcUint64 and funcInt64, too. I see that the bot also suggested it.

Also, I would rename funcInt --> funcInt32 and funcUint --> funcUint32 for consistency.

Then I'd reorder these as follows:

```slang
void funcInt8(int8_t x) {}
void funcInt16(int16_t x) {}
void funcInt32(int x) {}
void funcInt64(int64_t x) {}
void funcUint8(uint8_t x) {}
void funcUint16(uint16_t x) {}
void funcUint32(uint x) {}
void funcUint64(uint64_t x) {}
void funcHalf(half x) {}
```
or similar. Makes it easier to keep track of the definitions.

The 64-bit functions overflowing the literals possibly give another error (something about overflowing the int literal). But I think it makes sense to check here, anyway.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    funcUint8(255);
+    funcInt16(32767);
+    funcInt16(-32768);
+    funcUint16(65535);
+    funcInt(42);
+    funcUint(42u);
+    funcHalf(2048);
+
+    // Bit-width boundary: max values that fit exactly in the type's bit width
+    funcInt8(255);
```
Comment (skiminki-nv) [inline tests/diagnostics/integer-constant-overflow.slang:112]: I'm not sure about this. Perhaps there's a way to do better here. Do we still have the info available whether the literal was given as a decimal or hex? I think that in hex or octal, we could give this a pass. But for a decimal literal, passing 255 for int8_t parameter is almost certainly an error.

If it turns out that we don't have the info about how the original format of the literal, then maybe file a bug and we'll clean up this later?

File: tests/diagnostics/integer-constant-overflow.slang [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    funcUint(4294967295u);
+
+    // Values outside the signed range but within the type's bit width do
+    // NOT trigger the overflow warning. The check is conservative (bit-width
+    // only, not range-aware) to avoid false positives on common shader
+    // patterns where hex constants serve as bit patterns
+    // (e.g. packed bytes: int x = 0xFF030206).
+    funcInt(0xFFFFFFFFu);
+    funcInt8(128);
+    funcInt8(-129);
```
Comment (skiminki-nv) [inline tests/diagnostics/integer-constant-overflow.slang]: This should trigger a warning. -129 is never representable by an 8-bit integer.

File: tests/diagnostics/integer-constant-overflow.slang (hunk 3 above)
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:112]: Filed issue #10516.

File: tests/diagnostics/integer-constant-overflow.slang (hunk 1 above)
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:101]: added funcHalf(131072) with annotated warning.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    // reduces cost below kConversionCost_Default, so no warning from either check.
+    funcUint(-2);
+
+    // -1 to unsigned is allowed (common "all ones" pattern)
+    funcUint(-1);
+    funcUint8(-1);
+
+    // Explicit casts should NOT warn (handled as ExplicitCoercion)
+    int8_t y = int8_t(256);
+}
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:138]: Added int8_t returnOverflow()

File: tests/diagnostics/integer-constant-overflow.slang [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    // Variable initializer overflow
+    int8_t x = 256;
+//CHECK:       ^^^ integer constant overflow in conversion
+//CHECK:       ^^^ integer value '256' does not fit in type 'int8_t'
+
+    // Constant expressions that overflow
+    funcInt8(100 + 200);
+//CHECK:         ^ integer constant overflow in conversion
+//CHECK:         ^ integer value '300' does not fit in type 'int8_t'
+
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:62]: Done

File: tests/diagnostics/integer-constant-overflow.slang (hunk 2 above)
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:11]: Done, including 64-bit overflow cases

File: source/slang/slang-check-conversion.cpp [hunk 7]
```diff
@@ -2027,6 +2027,51 @@ bool SemanticsVisitor::_coerce(
...
+            {
+                if (!hasFolded)
+                {
+                    cachedFoldedVal = tryFoldIntegerConstantExpression(
+                        fromExpr,
+                        ConstantFoldingKind::CompileTime,
+                        nullptr);
+                    hasFolded = true;
+                }
+                return as<ConstantIntVal>(cachedFoldedVal);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Addressed — now caching as `ConstantIntVal*` directly.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    // Values outside the signed range but within the type's bit width do
+    // NOT trigger the overflow warning. The check is conservative (bit-width
+    // only, not range-aware) to avoid false positives on common shader
+    // patterns where hex constants serve as bit patterns
+    // (e.g. packed bytes: int x = 0xFF030206).
+    funcInt(0xFFFFFFFFu);
+    funcInt8(128);
+    funcInt8(-129);
+    funcInt8(-200);
+    funcInt16(40000);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang]: Addressed — `funcInt8(-129)` and `funcInt8(-200)` now correctly trigger the overflow warning. The negative-value branch uses a range-based check (`|v| > 2^(n-1)`) that catches these.

File: tests/diagnostics/integer-constant-overflow.slang (hunk 4 above)
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang]: Addressed in cab643c — now correctly detects -129, -200, and -256 as overflow for int8_t using the range-based negative check.

File: source/slang/slang-check-conversion.cpp [hunk 9]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
...
+                }
+                return cachedFoldedVal;
+            };
+
+            int maxBitSize = getMaximumTypeBitSize(toType);
+            if (!isCoreModule && maxBitSize > 0 && cost < kConversionCost_Explicit)
+            {
+                if (auto val = getFoldedIntVal())
+                {
+                    IntegerLiteralValue v = val->getValue();
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2117]: Fixed in 289338d — using unsigned subtraction (`0ULL - uint64_t(v)`) instead of signed negation to avoid UB when v == INT64_MIN.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    // reduces cost below kConversionCost_Default, so no warning from either check.
+    funcUint32(-2);
+
+    // -1 to unsigned is allowed (common "all ones" pattern)
+    funcUint32(-1);
+    funcUint8(-1);
+
+    // Explicit casts should NOT warn (handled as ExplicitCoercion)
+    int8_t y = int8_t(256);
+}
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:138]: Good suggestions. These are out of scope for this PR — vector types, enum types, and struct fields involve different code paths. Could be addressed in a follow-up.

File: source/slang/slang-check-conversion.cpp [hunk 11]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
         //
         if (outToExpr && site != CoercionSite::ExplicitCoercion)
         {
+            bool overflowWarningDetected = false;
+            bool isCoreModule = false;
+            if (auto module = getShared()->getModule())
+                if (auto moduleDecl = module->getModuleDecl())
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2092]: Acceptable as-is — the `isCoreModule` check is two cheap pointer dereferences and a modifier lookup. Moving it inside the `maxBitSize > 0` guard would save negligible work for non-integer types while reducing readability.

File: source/slang/slang-check-conversion.cpp [hunk 12]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
...
+            };
+
+            int maxBitSize = getMaximumTypeBitSize(toType);
+            if (!isCoreModule && maxBitSize > 0 && cost < kConversionCost_Explicit)
+            {
+                if (auto val = getFoldedIntVal())
+                {
+                    IntegerLiteralValue v = val->getValue();
+                    bool overflow = false;
+                    if (v < 0)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2119]: Good point. The code comments in the test file already explain the rationale for each branch. The asymmetry is intentional and documented in the PR description and test comments.

File: source/slang/slang-check-conversion.cpp [hunk 13]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
...
+            bool isCoreModule = false;
+            if (auto module = getShared()->getModule())
+                if (auto moduleDecl = module->getModuleDecl())
+                    isCoreModule = moduleDecl->hasModifier<FromCoreModuleModifier>();
+
+            // Cache the constant-fold result so both the overflow check and
+            // the UnrecommendedImplicitConversion check can reuse it.
+            ConstantIntVal* cachedFoldedVal = nullptr;
+            bool hasFolded = false;
+            auto getFoldedIntVal = [&]() -> ConstantIntVal*
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2099]: Thanks!

File: tests/diagnostics/integer-constant-overflow.slang [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    // where hex constants serve as bit patterns
+    // (e.g. packed bytes: int x = 0xFF030206).
+    funcInt32(0xFFFFFFFFu);
+    funcInt8(128);
+    funcInt16(40000);
+
+    // -2 to unsigned: bit-width fits and getImplicitConversionCostWithKnownArg
+    // reduces cost below kConversionCost_Default, so no warning from either check.
+    funcUint32(-2);
+
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:131]: The negative overflow path is already exercised by `funcInt8(-129)`, `funcInt8(-200)`, and `funcInt8(-256)`. The unsigned variant would just confirm the same `|v| > 2^(n-1)` check fires for a different target type — marginal coverage gain.

File: source/slang/slang-check-conversion.cpp [hunk 15]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
...
+                    if (v < 0)
+                        overflow = (0ULL - static_cast<uint64_t>(v)) >
+                                   (static_cast<uint64_t>(1) << (maxBitSize - 1));
+                    else
+                        overflow = getIntValueBitSize(v) > maxBitSize;
+
+                    if (overflow)
+                    {
+                        // Set even when sink is null (probe/costing calls) to
+                        // consistently suppress the UnrecommendedImplicitConversion
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Acknowledged — the flag is set conservatively for future-proofing. No functional impact as noted.

File: tests/diagnostics/integer-constant-overflow.slang (hunk 4 above)
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang]: Fixed in commit cab643c. The negative overflow detection now uses a range-aware check (`|v| > 2^(N-1)`) that correctly catches -129, -200, and -256 for int8_t.

File: source/slang/slang-check-conversion.cpp [hunk 16]
```diff
@@ -2027,6 +2027,35 @@ bool SemanticsVisitor::_coerce(
...
+                            }
+                            overflowWarningEmitted = true;
+                        }
+                    }
+                }
+            }
+
             if (cost >= kConversionCost_Explicit)
             {
                 if (sink)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Addressed - the overflow check is now guarded with `cost < kConversionCost_Explicit` to avoid double-diagnosing.

File: source/slang/slang-check-conversion.cpp [hunk 17]
```diff
@@ -2027,6 +2027,36 @@ bool SemanticsVisitor::_coerce(
...
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2155]: Addressed - added the `cost < kConversionCost_Explicit` guard.

File: source/slang/slang-check-conversion.cpp [hunk 18]
```diff
@@ -2027,6 +2027,36 @@ bool SemanticsVisitor::_coerce(
...
+                            if (sink)
+                            {
+                                sink->diagnose(Diagnostics::IntegerConstantOverflow{
+                                    .value = String(val->getValue()),
+                                    .toType = toType,
+                                    .expr = fromExpr});
+                            }
+                            overflowWarningEmitted = true;
+                        }
+                    }
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Addressed - the fold result is now cached via the `getFoldedIntVal` lambda so both checks reuse the same result.

File: source/slang/slang-check-conversion.cpp [hunk 19]
```diff
@@ -2027,6 +2027,36 @@ bool SemanticsVisitor::_coerce(
...
+                        ConstantFoldingKind::CompileTime,
+                        nullptr))
+                {
+                    if (auto val = as<ConstantIntVal>(intVal))
+                    {
+                        if (getIntValueBitSize(val->getValue()) > maxBitSize)
+                        {
+                            if (sink)
+                            {
+                                sink->diagnose(Diagnostics::IntegerConstantOverflow{
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Addressed - renamed to `overflowWarningDetected` and set unconditionally (outside the `if(sink)` block) as suggested.

File: source/slang/slang-check-conversion.cpp [hunk 20]
```diff
@@ -2041,10 +2071,8 @@ bool SemanticsVisitor::_coerce(
                         .location = fromExpr->loc});
                 }
             }
-            else if (cost >= kConversionCost_Default)
+            else if (cost >= kConversionCost_Default && !overflowWarningEmitted)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Addressed - updated the comment to explain both the original intent and the new overflow suppression logic.

File: source/slang/slang-check-conversion.cpp [hunk 21]
```diff
@@ -2027,6 +2027,48 @@ bool SemanticsVisitor::_coerce(
...
+            {
+                if (auto val = getFoldedIntVal())
+                {
+                    if (getIntValueBitSize(val->getValue()) > maxBitSize)
+                    {
+                        if (sink)
+                        {
+                            sink->diagnose(Diagnostics::IntegerConstantOverflow{
+                                .value = String(val->getValue()),
+                                .toType = toType,
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2146]: Addressed - renamed to `overflowWarningDetected` and set unconditionally as suggested.

File: source/slang/slang-check-conversion.cpp [hunk 22]
```diff
@@ -2041,25 +2083,19 @@ bool SemanticsVisitor::_coerce(
                         .location = fromExpr->loc});
                 }
             }
-            else if (cost >= kConversionCost_Default)
+            // For general implicit conversions with high cost, emit a warning
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2167]: Addressed - the replacement comment now preserves the original explanation.

File: source/slang/slang-check-conversion.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Addressed - the lambda now caches `ConstantIntVal*` directly instead of `Val*`.

File: source/slang/slang-check-conversion.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2117]: Fixed in commit 289338d. Using `0ULL - static_cast<uint64_t>(v)` to avoid both the signed overflow UB and the MSVC warning.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+    funcUint8(-1);
+
+    // Explicit casts should NOT warn (handled as ExplicitCoercion)
+    int8_t y = int8_t(256);
+
+    // Constant expressions that overflow
+    funcInt8(100 + 200);
+//CHECK:         ^ integer constant overflow in conversion
+//CHECK:         ^ integer value '300' does not fit in type 'int8_t'
+}
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:138]: Addressed - funcInt8(128) and funcInt8(-129) are now covered. -129 triggers a warning (negative range-aware check). 128 intentionally does not warn (bit-width design choice, documented in test).

File: tests/diagnostics/integer-constant-overflow.slang [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+    funcUint(42u);
+    funcHalf(2048);
+
+    // -1 to unsigned is allowed (common "all ones" pattern)
+    funcUint(-1);
+    funcUint8(-1);
+
+    // Explicit casts should NOT warn (handled as ExplicitCoercion)
+    int8_t y = int8_t(256);
+}
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:138]: Addressed - added uint16_t overflow (funcUint16(65536)) and 32-bit boundary cases (funcInt32(4294967296), funcUint32(4294967296)).

File: tests/diagnostics/integer-constant-overflow.slang [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+    funcInt16(40000);
+    funcUint(-2);
+
+    // -1 to unsigned is allowed (common "all ones" pattern)
+    funcUint(-1);
+    funcUint8(-1);
+
+    // Explicit casts should NOT warn (handled as ExplicitCoercion)
+    int8_t y = int8_t(256);
+}
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang]: Correct - funcInt8(128) is intentionally silent from both checks. The bit-width check doesn't flag it (8 bits fits in 8), and the cost is reduced below kConversionCost_Default so UnrecommendedImplicitConversion also doesn't fire. This is the deliberate design choice documented in the test comments.

File: source/slang/slang-check-conversion.cpp [hunk 26]
```diff
@@ -2027,6 +2027,48 @@ bool SemanticsVisitor::_coerce(
...
+                        ConstantFoldingKind::CompileTime,
+                        nullptr);
+                    hasFolded = true;
+                }
+                return as<ConstantIntVal>(cachedFoldedVal);
+            };
+
+            int maxBitSize = getMaximumTypeBitSize(toType);
+            if (!isCoreModule && maxBitSize > 0 && cost < kConversionCost_Explicit)
+            {
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2114]: Good catch on the description inaccuracy. The current code uses bit-width comparison for positive values and range-aware comparison for negative values. The PR description has been updated to reflect this.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang]: Updated - funcInt8(-129) and funcInt8(-200) now correctly trigger the warning after adding negative range-aware overflow detection. The test comments have been clarified.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+    // 4294967296 (2^32) needs 33 bits, exceeding int's 32
+    funcInt(4294967296);
+//CHECK:    ^^^^^^^^^^ integer constant overflow in conversion
+//CHECK:    ^^^^^^^^^^ integer value '4294967296' does not fit in type 'int'
+
+    // Function arguments that fit - should NOT warn
+    funcInt8(127);
+    funcInt8(-128);
+    funcUint8(255);
+    funcInt16(32767);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:94]: Addressed - added funcInt8(255), funcUint32(4294967295u), and funcUint32(4294967296) boundary test cases.

File: source/slang/slang-check-conversion.cpp [hunk 29]
```diff
@@ -2027,6 +2027,48 @@ bool SemanticsVisitor::_coerce(
...
+                if (auto val = getFoldedIntVal())
+                {
+                    if (getIntValueBitSize(val->getValue()) > maxBitSize)
+                    {
+                        overflowWarningDetected = true;
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2141]: Thanks. The `overflowWarningDetected` comment has been improved to clarify that it's set for consistent suppression regardless of sink state.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+[numthreads(1, 1, 1)]
+void computeMain()
+{
+    // Generic value argument overflow
+    GenericInt8<256> g;
+//CHECK:        ^^^ integer constant overflow in conversion
+//CHECK:        ^^^ integer value '256' does not fit in type 'int8_t'
+
+    // Function arguments that overflow - should warn
+    funcInt8(256);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:30]: Good point - added a comment at the top of the warn section documenting the duplicate-suppression behavior.

File: source/slang/slang-check-conversion.cpp [hunk 31]
```diff
@@ -2027,6 +2027,51 @@ bool SemanticsVisitor::_coerce(
...
+                return as<ConstantIntVal>(cachedFoldedVal);
+            };
+
+            int maxBitSize = getMaximumTypeBitSize(toType);
+            if (!isCoreModule && maxBitSize > 0 && cost < kConversionCost_Explicit)
+            {
+                if (auto val = getFoldedIntVal())
+                {
+                    if (getIntValueBitSize(val->getValue()) > maxBitSize)
+                    {
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2137]: Correct - that's the intended behavior. When cost is 600 and overflow is detected, the user gets the more specific overflow warning and the general warning is suppressed via `overflowWarningDetected`.

File: tests/diagnostics/integer-constant-overflow.slang (hunk 8 above)
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang]: Thanks for the observation. funcInt8(-129) and -200 now correctly trigger the overflow warning after adding negative range detection. The interplay is: negative values with magnitude > 2^(N-1) trigger the overflow warning, which suppresses UnrecommendedImplicitConversion.

File: source/slang/slang-check-conversion.cpp [hunk 32]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
...
+                if (auto val = getFoldedIntVal())
+                {
+                    IntegerLiteralValue v = val->getValue();
+                    bool overflow = false;
+                    if (v < 0)
+                        overflow = (0ULL - static_cast<uint64_t>(v)) >
+                                   (static_cast<uint64_t>(1) << (maxBitSize - 1));
+                    else
+                        overflow = getIntValueBitSize(v) > maxBitSize;
+
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2135]: Addressed - updated the comment wording as suggested.

File: source/slang/slang-check-conversion.cpp [hunk 33]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
...
+            int maxBitSize = getMaximumTypeBitSize(toType);
+            if (!isCoreModule && maxBitSize > 0 && cost < kConversionCost_Explicit)
+            {
+                if (auto val = getFoldedIntVal())
+                {
+                    IntegerLiteralValue v = val->getValue();
+                    bool overflow = false;
+                    if (v < 0)
+                        overflow = (0ULL - static_cast<uint64_t>(v)) >
+                                   (static_cast<uint64_t>(1) << (maxBitSize - 1));
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Addressed - added an inline comment explaining the asymmetry between the positive (bit-width) and negative (magnitude > 2^(N-1)) paths.

File: source/slang/slang-check-conversion.cpp [hunk 34]
```diff
@@ -2027,6 +2027,59 @@ bool SemanticsVisitor::_coerce(
...
+            auto getFoldedIntVal = [&]() -> ConstantIntVal*
+            {
+                if (!hasFolded)
+                {
+                    cachedFoldedVal = as<ConstantIntVal>(tryFoldIntegerConstantExpression(
+                        fromExpr,
+                        ConstantFoldingKind::CompileTime,
+                        nullptr));
+                    hasFolded = true;
+                }
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2108]: Acknowledged - the earlier fold for low-cost conversions is intentional and shouldn't be a perf concern.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 35]
```diff
@@ -0,0 +1,128 @@
...
+    // do NOT trigger the overflow warning. The check is conservative (bit-width
+    // only, not range-aware) to avoid false positives on common shader patterns
+    // where hex constants serve as bit patterns
+    // (e.g. packed bytes: int x = 0xFF030206).
+    funcInt32(0xFFFFFFFFu);
+    funcInt8(128);
+    funcInt16(40000);
+
+    // -2 to unsigned: bit-width fits and getImplicitConversionCostWithKnownArg
+    // reduces cost below kConversionCost_Default, so no warning from either check.
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:129]: Addressed - added `funcUint8(-129)` test case.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 36]
```diff
@@ -0,0 +1,128 @@
...
+    // returns 0 for half). Values beyond 16-bit integers still don't trigger
+    // the integer overflow warning, but do trigger the general implicit
+    // conversion warning since they exceed half's exact integer range.
+    funcHalf(131072);
+//CHECK:     ^^^^^^ implicit conversion not recommended
+//CHECK:     ^^^^^^ implicit conversion from 'int' to 'half' is not recommended
+
+    // Bit-width boundary: max values that fit exactly in the type's bit width
+    funcInt8(255);
+    funcUint32(4294967295u);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:113]: Thanks!

File: source/slang/slang-check-conversion.cpp [hunk 37]
```diff
@@ -2027,6 +2027,63 @@ bool SemanticsVisitor::_coerce(
...
+                if (auto val = getFoldedIntVal())
+                {
+                    IntegerLiteralValue v = val->getValue();
+                    bool overflow = false;
+                    // For negative values, overflow if magnitude exceeds 2^(N-1)
+                    // (the two's complement minimum). For positive values, use
+                    // bit-width comparison to avoid false positives on hex
+                    // bit-pattern idioms like int x = 0xFF030206.
+                    if (v < 0)
+                        overflow = (0ULL - static_cast<uint64_t>(v)) >
```
Comment (skiminki-nv) [inline source/slang/slang-check-conversion.cpp]: This check looks more complicated than it should. At the minimum, this needs more commentary to explain why this works for INT64_MIN.

I would rather formulate this without manipulating the value of `v` at all if possible. That is, simply compute the minimum allowed negative value for maxBitSize and check whether `v` fits.

Possibly something like:

    int64_t minValue = static_cast<int64_t>(INT64_C(1) << (maxBitSize - 1));
    overflow = v < minValue;

or put the min value in a separate variable and

Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp]: Fixed in 4c75078. Replaced the unsigned magnitude arithmetic with a direct minimum-value comparison: compute \-(INT64_C(1) << (maxBitSize - 1))\ and check \ < minValue\. Guarded with \maxBitSize < 64\ since any \int64_t\ value fits in 64 bits, avoiding the UB concern entirely.

File: source/slang/slang-check-conversion.cpp [hunk 38]
```diff
@@ -2086,6 +2086,70 @@ bool SemanticsVisitor::_coerce(
         //
         if (outToExpr && site != CoercionSite::ExplicitCoercion)
         {
+            bool overflowWarningDetected = false;
+            bool isCoreModule = false;
+            if (auto module = getShared()->getModule())
+                if (auto moduleDecl = module->getModuleDecl())
+                    isCoreModule = moduleDecl->hasModifier<FromCoreModuleModifier>();
+
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2094]: Acknowledged. This is cheap enough to not matter in practice, so leaving as-is for now.

File: source/slang/slang-check-conversion.cpp [hunk 39]
```diff
@@ -2086,6 +2086,70 @@ bool SemanticsVisitor::_coerce(
...
+                        // Two's complement minimum for N bits is -(2^(N-1)).
+                        // For 64-bit targets, any int64_t value fits by definition.
+                        if (maxBitSize < 64)
+                        {
+                            int64_t minValue = -(INT64_C(1) << (maxBitSize - 1));
+                            overflow = v < minValue;
+                        }
+                    }
+                    else
+                    {
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/slang/slang-check-conversion.cpp:2130]: Added `funcUint8(-128)` as an explicit no-warn test case with a comment explaining the boundary behavior.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 40]
```diff
@@ -0,0 +1,131 @@
...
+
+    // Positive values outside the signed range but within the type's bit width
+    // do NOT trigger the overflow warning. The check is conservative (bit-width
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:116]: Added `funcUint8(-128)` with documenting comment.

File: tests/diagnostics/integer-constant-overflow.slang [hunk 41]
```diff
@@ -0,0 +1,131 @@
...
+//CHECK:       ^^^ integer value '-129' does not fit in type 'uint8_t'
+
+    // Constant expressions that overflow
+    funcInt8(100 + 200);
+//CHECK:         ^ integer constant overflow in conversion
+//CHECK:         ^ integer value '300' does not fit in type 'int8_t'
+
+    // 65536 needs 17 bits, exceeding 16-bit types
+    funcInt16(65536);
+//CHECK:      ^^^^^ integer constant overflow in conversion
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/integer-constant-overflow.slang:68]: Added `funcInt8(300 - 200)` as a no-warn case to verify the check evaluates the folded result.

**Comment by jvepsalainen-nv** [comment]
Replying to the asymmetry comment nit: the existing inline comments already explain the different strategies (negative: magnitude > 2^(N-1), positive: bit-width to avoid false positives on hex bit-pattern idioms). Keeping as-is.

Resolving all remaining bot review threads - they are positive observations or non-blocking suggestions.
